### PR TITLE
Upgrade to Karaf 4.4.3

### PR DESCRIPTION
**Pull request by wborn**
* Sync runtime dependencies with Karaf 4.4.3, most notably:
  * Jetty 9.4.50.v20221201
  * Pax Logging 2.2.0
  * Pax Web 8.0.15
* Use OSGi R8 as compile dependency
* Rework Servlets to use Http Whiteboard annotations in favor of proprietary `org.openhab.core.io.http.servlet` classes
* Resolve itest runbundles

---

To do:

* [x] Rebase and fix merge conflicts
* [x] Upgrade to the latest greatest Karaf 4.4 version
* [x] Upgrade dependencies to use OSGi R8

### CODE REVIEW

**Comment by J-N-K** [comment]
There are still issues. BasicUI starts, also charts and icons are available, but MainUI does not. 

WebSocket servlet is present, but fails to create the socket, I believe I'm running into the same issue as described here: https://github.com/eclipse/jetty.project/issues/7835.

**Comment by wborn** [comment]
I did have Main UI working with my changes but I can imagine with https://github.com/openhab/openhab-webui/pull/1521 it needs to be reworked. The main issue I had with main UI was that it didn't allow for any other servlets to be registered. So that's why I went with those Whiteboard changes which is also more standard compared to the proprietary OpenHABServlet classes etc. I've added you as collaborator on my webui repo so I think you can also push to my webui PR branch now.

**Comment by J-N-K** [comment]
MainUI works now (mostly, there is still one issue with serving images, but I guess that can be figured out.

One issue is that we declare `service = Foo.class` on some servlet components and then they are not registered. It is mandatory they also provide `Servlet.class`.

**Comment by J-N-K** [comment]
Regarding core, on startup I see:

```
Launching the openHAB runtime...
WARNING: package org.apache.karaf.specs.locator not in java.base
Listening for transport dt_socket at address: 5005
Error starting karaf activator org.apache.karaf.specs.activator.Activator: org/apache/karaf/specs/locator/OsgiLocator
```

That looks similar to https://issues.apache.org/jira/browse/KARAF-5840

But I'm not sure if this is similar to

**Comment by J-N-K** [comment]
This one looks pretty good now, apart from the itests that don't resolve. Maybe you can have a look at that, @wborn.

**Comment by J-N-K** [comment]
I fixed the tests (here and in addons). If you review my contributions, I'm fine with yours and we can merge if the web issues are resolved.

**Comment by J-N-K** [comment]
We'll have Karaf 4.4.3 probably tomorrow: https://lists.apache.org/list.html?dev@karaf.apache.org

**Comment by wborn** [comment]
If it works well that would be a nice latest greatest version to use with the PRs. :slightly_smiling_face:

File: bundles/org.openhab.core.io.websocket/src/main/java/org/openhab/core/io/websocket/package-info.java
Comment (wborn) [inline bundles/org.openhab.core.io.websocket/src/main/java/org/openhab/core/io/websocket/package-info.java:13]: Is this dynamic import really necessary?

Comment (J-N-K) [reply bundles/org.openhab.core.io.websocket/src/main/java/org/openhab/core/io/websocket/package-info.java:13]: Yes, even though I don't fully understand why. But I found that this is the solution for https://github.com/openhab/openhab-core/pull/3252#issuecomment-1374418213.

Comment (wborn) [reply bundles/org.openhab.core.io.websocket/src/main/java/org/openhab/core/io/websocket/package-info.java:13]: Did you also try adding an import for `org.eclipse.jetty.websocket.server.WebSocketServerFactory` in the code instead? That way bnd will add an import for the package to the manifest.

File: itests/org.openhab.core.model.script.tests/itest.bndrun
Comment (J-N-K) [inline itests/org.openhab.core.model.script.tests/itest.bndrun:8]: same here

Comment (wborn) [reply itests/org.openhab.core.model.script.tests/itest.bndrun:8]: I only changed the indentation from spaces to tabs. :slightly_smiling_face:

File: bundles/org.openhab.core.io.websocket/src/main/java/org/openhab/core/io/websocket/package-info.java
Comment (J-N-K) [reply bundles/org.openhab.core.io.websocket/src/main/java/org/openhab/core/io/websocket/package-info.java]: That works, too. Why is that superior to the dynamic-import?

Comment (wborn) [reply bundles/org.openhab.core.io.websocket/src/main/java/org/openhab/core/io/websocket/package-info.java]: It will make sure that the required package imports are satisfied before the bundle is started so you'll know well in advance that something is wrong. That can be long before the code is executed that looks up the class.

**Comment by wborn** [comment]
Thank you very much for your help @J-N-K. :+1: 
I think we are done with this so I think it is time for the other maintainers to review this.

**Comment by wborn** [comment]
> Do we need to check zwave and zigbee?

AFAIK they don't use servlets so I think they are fine. :slightly_smiling_face:

File: bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/internal/AudioServlet.java
Comment (J-N-K) [inline bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/internal/AudioServlet.java:56]: This is more a comment about code-style (and not necessarily to be enforced here): There are dozens of different arrangements for the annotations, sometime `@Component` is first and `@NonNullByDefault` is at the end, sometime it's exactly the opposite and sometime they are together with the whiteboard annotations above or below. 

Can we decide for one convention and probably enable https://github.com/checkstyle/checkstyle/issues/11693 if it is implemented in the customizable way someday? I would agree with https://stackoverflow.com/a/5638388 and order by importance, i.e. `@Component`, other and `@NonNullByDefault` at the end as it is present on nearly every class.

Comment (wborn) [reply bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/internal/AudioServlet.java:56]: I usually sort them which most of the time also goes well with the order of importance.

**Review by J-N-K — APPROVED** [review]

**Review by J-N-K — APPROVED** [review]

**Comment by wborn** [comment]
I will resolve the merge conflicts after the review is done. :wink:

**Review by kaikreuzer — APPROVED** [review]
Thanks for the work @wborn!
I have only glanced over the code and not tested it, but this is anyhow tricky - but what I see lgtm.

> I will resolve the merge conflicts after the review is done. 😉

It looks to me as if @J-N-K already reviewed and approved it 2 weeks ago - if you waited for me to approve as well: Here you go!

**Comment by J-N-K** [comment]
@kaikreuzer we did wait for you because I contributed to the code 😀

**Comment by florian-h05** [comment]
It seems that the developer docs are outdated: https://next.openhab.org/docs/developer/guidelines.html#d-language-levels-and-libraries
The say that openHAB is using OSGi R7, but isn’t it using R8 now?

**Pull request merged**
